### PR TITLE
Add Git for spark notebook community

### PR DESCRIPTION
**Pull request by vidma**
depends on https://github.com/kensuio/spark-notebook/pull/1 (which adds git-notebook-provider module as is).

- [x] **already working with HTTPS**
one needs to add env vars to his `~/.bashrc` and run `./run-dev.sh git`
```
export SN_NOTEBOOKS_GIT_HTTPS_REPO="https://github.com/vidma/snb-tests.git"
export SN_NOTEBOOKS_GIT_USER="floatvilnius"
export SN_NOTEBOOKS_GIT_TOKEN="token"
```
- [ ] **ssh key do not seem to work yet**, I think we can ship without if it won't be easy to fix.
- [ ] still quite nasty if git config is incorrect. we may improve at least error handling/checking here. e.g. check if credentials are correct for given repo url
- [x] works with SNC 1.6 and SNC 2.0

### CODE REVIEW

File: app/controllers/Application.scala [hunk 1]
```diff
@@ -500,24 +482,27 @@ object Application extends Controller {
...
+      Json.obj("id" -> id, "message" -> message, "last_modified" -> ts)
+    }
+    Ok(JsArray(cs))
   }
 
+  def restoreCheckpoint(snb:String, id:String) = Action { request =>
+    //TODO → retrieve checkpoint and overwritte the notebook locally (until next checkpoint)
+    val path = URLDecoder.decode(snb, UTF_8)
+    notebookManager.restoreCheckpoint(path, id)
+    println(s" #### ID passed by restore: [$id]")
```
Comment (maasg) [inline app/controllers/Application.scala]: debug output needed?

File: app/notebook/server/NotebookManager.scala [hunk 2]
```diff
@@ -85,82 +98,104 @@ class NotebookManager(val name: String, val notebookDir: File) {
...
+    } else {
+        Failure(new DirectoryCreationException(newDir))
+    }
+  }
+
   def getNotebook(path: String) = {
     Logger.info(s"getNotebook at path $path")
     for (notebook <- load(path)) yield {
       val data = FileUtils.readFileToString(notebookFile(path), "UTF-8")
       val df = new SimpleDateFormat("dd-MM-yyyy HH:mm:ss z'('Z')'")
```
Comment (maasg) [inline app/notebook/server/NotebookManager.scala]: do we support java 7 ? Otherwise we could migrate the dangerous* `SimpleDateFormat` to `java.time`
(* probably not in this specific context)

File: app/notebook/server/NotebookManager.scala [hunk 3]
```diff
@@ -85,82 +98,104 @@ class NotebookManager(val name: String, val notebookDir: File) {
...
       val last_mtime = df.format(new Date(notebookFile(path).lastModified()))
```
Comment (maasg) [inline app/notebook/server/NotebookManager.scala]: `last_mtime` -> `lastModifiedTime` ?

File: app/notebook/server/NotebookManager.scala [hunk 4]
```diff
@@ -85,82 +98,104 @@ class NotebookManager(val name: String, val notebookDir: File) {
...
+
+  def checkpoints(path: String): List[Version] = {
+    val file = notebookFile(path)
+    Await.result(provider.versions(Paths.get(file.toURI)), DefaultOperationTimeout)
+  }
+
+  def restoreCheckpoint(path:String, id:String): Option[(String, String, String)] = {
+    val absolutePath = Paths.get(notebookFile(path).toURI)
+    val notebookCheckpoint =  provider.get(absolutePath, Some(Version(id,"", 1L))).map{nb =>
+      val df = new SimpleDateFormat("dd-MM-yyyy HH:mm:ss z'('Z')'")
```
Comment (maasg) [inline app/notebook/server/NotebookManager.scala]: maybe we could centralize the date pattern we use to avoid consistency issues?

File: app/notebook/server/NotebookManager.scala [hunk 5]
```diff
@@ -85,82 +98,104 @@ class NotebookManager(val name: String, val notebookDir: File) {
...
 
   def load(path: String): Option[Notebook] = {
     Logger.info(s"Loading notebook at path $path")
-    val file = notebookFile(path)
-    if (file.exists())
-      Notebook.deserialize(FileUtils.readFileToString(file))
-    else
-      None
+    // TODO: this is expected to be sync, and the implementation is ugly...
+    Await.result( provider.get( Paths.get(notebookFile(path).toURI) ).map(Option.apply), DefaultOperationTimeout ).orElse( None )
```
Comment (maasg) [inline app/notebook/server/NotebookManager.scala]: there's a lot going on on this line. Maybe split to make it more digestable?  
In particular, we use this `x => Paths.get(x).toURI` several times. Maybe it's worth encapsulating.

File: build.sbt [hunk 6]
```diff
@@ -231,9 +231,21 @@ lazy val sparkNotebookCore = Project(id = "spark-notebook-core", base = file("mo
     Extra.sparkNotebookCoreSettings
   )
 
+lazy val gitNotebookProvider = Project(id = "git-notebook-provider", base = file("modules/git-notebook-provider"))
+  .settings(
+    scalaVersion := defaultScalaVersion,
+    organization := "guru.data-fellas",
```
Comment (maasg) [inline build.sbt]: should we change to `io.kensu` ?

File: modules/core/src/main/scala/notebook/Notebook.scala [hunk 7]
```diff
@@ -1,20 +1,23 @@
 package notebook
 
+import scala.concurrent.Future
+import scala.concurrent.ExecutionContext
 import notebook.NBSerializer._
 
 case class Notebook(
                      metadata: Option[Metadata] = None,
```
Comment (maasg) [inline modules/core/src/main/scala/notebook/Notebook.scala]: (side note to this PR: is `metadata` really optional?)

File: modules/core/src/main/scala/notebook/Notebook.scala [hunk 8]
```diff
@@ -23,4 +26,24 @@ object Notebook {
   def deserialize(str: String): Option[Notebook] = NBSerializer.fromJson(str)
 
   def serialize(nb: Notebook): String = NBSerializer.toJson(nb)
+
+  def deserializeFuture(str: String)(implicit ex : ExecutionContext): Future[Notebook] = {
+    Future(deserialize(str).get)
```
Comment (maasg) [inline modules/core/src/main/scala/notebook/Notebook.scala]: This looks weird, as the only probable failure will be `java.util.NoSuchElementException: None.get` 
I think that `deserialize` should be a `Try[Notebook]`, reporting whether  it had an issue  in the deserialization process. Then `deserializeFuture` could be `Future.fromTry(tryDeserialize)`

File: project/Shared.scala [hunk 9]
```diff
@@ -17,6 +17,9 @@ object Shared {
 
   lazy val sharedSettings: Seq[Def.Setting[_]] = Seq(
     publishArtifact in Test := false,
+    publishMavenStyle := true,
+
+    organization := "guru.data-fellas",
```
Comment (maasg) [inline project/Shared.scala]: use `io.kensu` ?

File: public/ipython/notebook/js/notebook.js [hunk 10]
```diff
@@ -1904,6 +1908,34 @@ define([
...
+        } else {
+            this.list_checkpoints_timer = null;
+            //this.events.trigger("autosave_disabled.Notebook");
+        }
+    };
+
+    /**
+     * Start an autosave timer which periodically saves the notebook.
+     *
+     * @param {integer} interval - the autosave interval in milliseconds
```
Comment (maasg) [inline public/ipython/notebook/js/notebook.js]: Add a note about how interval is used?
"the autosave interval in milliseconds - 0 disables autosave"

File: public/ipython/notebook/js/notebook.js [hunk 11]
```diff
@@ -2413,6 +2457,7 @@ define([
      * @param {object} data - JSON representation of a checkpoint
      */
     Notebook.prototype.list_checkpoints_success = function (data) {
+        _.each(data, function(c) {c.last_modified = c.last_modified * 1000;});
```
Comment (maasg) [inline public/ipython/notebook/js/notebook.js]: don't understand the magic behind the `*1000` mutation. Is there a chance this happens more than once? (resulting in `*1000*1000*1000` ? )

File: build.sbt (hunk 6 above)
Comment (vidma) [inline build.sbt]: then we should change everywhere. that should be a separate PR.

File: public/ipython/notebook/js/notebook.js (hunk 11 above)
Comment (vidma) [inline public/ipython/notebook/js/notebook.js]: this is an ajax callback (at `notebook_path + '/checkpoints'` ) getting ajax result as `data` param -so this simply transform timestamps from seconds to milliseconds (or alike).
This transformation could possibly happen in backend too...

File: modules/core/src/main/scala/notebook/Notebook.scala (hunk 8 above)
Comment (vidma) [inline modules/core/src/main/scala/notebook/Notebook.scala]: 🆗 , but this changes API and is also not related to this PR/Git.

File: app/notebook/server/NotebookManager.scala (hunk 2 above)
Comment (vidma) [inline app/notebook/server/NotebookManager.scala]: see no reason to drop java7 support yet. will extract a helper though.

File: build.sbt (hunk 6 above)
Comment (vidma) [inline build.sbt]: made a separate PR for this https://github.com/kensuio/spark-notebook/pull/3 (quite a nasty change, and unrelated to this PR)

File: modules/core/src/main/scala/notebook/Notebook.scala (hunk 7 above)
Comment (vidma) [inline modules/core/src/main/scala/notebook/Notebook.scala]: not worth the API change IMHO.

File: modules/core/src/main/scala/notebook/Notebook.scala (hunk 8 above)
Comment (maasg) [inline modules/core/src/main/scala/notebook/Notebook.scala]: ok
